modules/vmss/main.tf

```diff
@@ -1,259 +1,370 @@
-  upgrade_mode = "Manual"
-
-  custom_data = base64encode(var.bootstrap_options)
+  source_image_reference {
+    publisher = var.image.custom_id == null ? var.image.publisher : null
+    offer     = var.image.custom_id == null ? var.image.offer : null
+    sku       = var.image.custom_id == null ? var.image.sku : null
+    version   = var.image.version
+  }
+
+  source_image_id = var.image.custom_id
+  os_disk {
+    caching                = "ReadWrite"
+    disk_encryption_set_id = var.virtual_machine_scale_set.disk_encryption_set_id #  The Disk Encryption Set must have the Reader Role Assignment scoped on the Key Vault - in addition to an Access Policy to the Key Vault.
+    storage_account_type   = var.virtual_machine_scale_set.disk_type
+  }
+
+
+  instances = var.autoscaling_configuration.default_count
+
+  upgrade_mode = "Manual" # See README for more details no this setting.
+
+  custom_data = var.virtual_machine_scale_set.bootstrap_options == null ? null : base64encode(var.virtual_machine_scale_set.bootstrap_options)
 
   scale_in {
-    rule                   = var.scale_in_policy
-    force_deletion_enabled = var.scale_in_force_deletion
+    rule                   = var.autoscaling_configuration.scale_in_policy
+    force_deletion_enabled = var.autoscaling_configuration.scale_in_force_deletion
   }
 
   dynamic "network_interface" {
     for_each = var.interfaces
     iterator = nic
 
     content {
-      name                          = "${var.name}-${nic.value.name}"
+      name                          = nic.value.name
       primary                       = nic.key == 0 ? true : false
       enable_ip_forwarding          = nic.key == 0 ? false : true
-      enable_accelerated_networking = nic.key == 0 ? false : var.accelerated_networking
+      enable_accelerated_networking = nic.key == 0 ? false : var.virtual_machine_scale_set.accelerated_networking
 
       ip_configuration {
         name                                         = "primary"
         primary                                      = true
         subnet_id                                    = nic.value.subnet_id
-        load_balancer_backend_address_pool_ids       = nic.key == 0 ? [] : try(nic.value.lb_backend_pool_ids, [])
-        application_gateway_backend_address_pool_ids = nic.key == 0 ? [] : try(nic.value.appgw_backend_pool_ids, [])
+        load_balancer_backend_address_pool_ids       = nic.value.lb_backend_pool_ids
+        application_gateway_backend_address_pool_ids = nic.value.appgw_backend_pool_ids
 
         dynamic "public_ip_address" {
-          for_each = try(nic.value.create_pip, false) ? ["one"] : []
+          for_each = nic.value.create_public_ip ? [1] : []
           iterator = pip
 
           content {
-            name              = "${var.name}-${nic.value.name}-pip"
-            domain_name_label = try(nic.value.pip_domain_name_label, null)
+            name              = "${nic.value.name}-public-ip"
```

Comment:
Can we avoid hardcoding fragments of names for PIP (`"${nic.value.name}-public-ip"`)?

Reply:
this name is not seen anywhere, you have to dig deep to find it. We have to give it some name, but it does not really matter which one 

it can be just the `nic.value.name`, that's fine with me